src/AutoRest.CSharp/Output/Models/Types/OutputLibrary.cs

```diff
@@ -68,6 +83,99 @@ public OutputLibrary(CodeModel codeModel, BuildContext context)
 
         public IEnumerable<ResponseHeaderGroupType> HeaderModels => (_headerModels ??= EnsureHeaderModels()).Values;
 
+        private static IList<System.Type> GetReferenceClassCollection()
```

Comment:
Lets extract the following methods into a class called InheritanceChoser

GetReferenceClassCollection
RebuildExactModelInheritance
GetExactMatch
IsEqual
IsAssignable
RebuildModelInheritance